### ISSUE TITLE
Validating proper usage of projections seems impossible

### ISSUE THREAD

**Issue by brian-mann** [ClickHouse/ClickHouse#71843]
**Company or project name**

Cypress.io

**What we're solving for**
A deterministic way to validate that queries have been written such that they **would** use projections that have been created on various tables

Why we want to:
- CH will not use projections where queries use the `final` modifier
- We have the `final` modifier "on by default" since we only query aggregated tables 
- In order to use projections, we need to ensure that when we query those tables we specifically turn off final with `SETTINGS final=0` in the subqueries that these use
- We want our automated tests to catch/ensure these settings have been properly written for all queries

**Describe the unexpected behaviour**
- Clickhouse provides query level settings easily validate that a query used a projection such as: [`force_optimize_projection`](https://clickhouse.com/docs/en/operations/settings/settings#force_optimize_projection) and [`force_optimize_projection_name `](https://clickhouse.com/docs/en/operations/settings/settings#force_optimize_projection_name)
- These work "as-advertised" - the problem is that from our internal tests, it seems that clickhouse will **not use projections** solely due to the amount of rows in a table. Normally in production we have more than enough rows to always warrant using the projections, but in our automated test environment, we don't have enough data to cause it to use the projection.
- The problem is that there is no way to actually utilize these settings (or any other form of validation) to ensure the queries **would** use the projections.
- Because of the inconsistent behavior of `force_optimize_projection`, there is no way to actually use this setting nor is there any way to actually validate that queries have been written correctly to even know if it **would** be able to use the projection
- There is no way to tell **the difference** between a query that has been misconfigured from being able to use a projection vs one that **is configured properly** but that won't use the projection due to the amount of data in the table. CH won't use a projection for two very different reasons (1) it doesn't have enough data or (2) the query is written wrong. Of those 2 reasons, one is a programmer error, the other is expected. Since you can't tell the difference, there is no way to actually catch that problem and react to it.

**Expected behavior**
- I should be able to turn `force_optimize_projection` on and get deterministic failures to know the difference between invalid queries (that wont **ever** use the projection) vs valid queries (that would use them)
- `force_optimize_projection` should behave consistently just like [`force_primary_key`](https://clickhouse.com/docs/en/operations/settings/settings#force_primary_key) (which is an ideal setting) 
- With `force_primary_key` on - we can automatically ensure that **all queries** will perform well in production and fail any test that doesn't use indexes
- Unlike `force_optimize_projection`, `force_primary_key` works so well because its deterministic and not contingent on the amount of data you have in your table. Whether you have 1 row or 1 trillion rows, it will error the same way.
- `force_optimize_projection` should behave similarly in a deterministic way that isn't contingent on the amount of data in the table. 
- As it stands, it's impossible to know **why** clickhouse didn't use a projection. We actually don't care (in our tests) whether it did use the projection... we want to validate that the query was configured properly such that **it would have** used the projection (if there was enough data in the table).
- Alternatively, if there was a way to force CH to use the projection - that would also work just fine. You would think that [`preferred_optimize_projection_name`](https://clickhouse.com/docs/en/operations/settings/settings#preferred_optimize_projection_name) achieves this, but alas it does not. That only works in the case where there are multiple projections on the table and multiple could be utilized. It doesn't force CH to use the projection if it decides it doesn't need to in the first place. While this is a separate issue, it's behavior is correlated to the overall problem and suffers from the same inconsistency.

**Extra Context**
I'm using the same query, the only difference is that in the first query I've truncated the table and inserted 1 row. In the second example I've inserted several hundred thousand rows.

```sql
select frame_view, frame_view_id from global_frame_view_identifiers_agg where frame_view in (...)
SETTINGS final=0,force_optimize_projection=1,force_optimize_projection_name='frame_view_by_frame_view_id_projection';
```

When doing `explain indexes=1` you can see that it uses the projection when there are many rows, and doesn't use the projection if there is only one row... therefore if you use `force_optimize_projection` it will sometimes error, and sometimes not error.

One Row:
![Screenshot 2024-11-12 at 12 59 25 PM copy](https://github.com/user-attachments/assets/0ad48d6d-6122-4e15-b87d-a2b80160dac5)

Thousands of Rows:
![Screenshot 2024-11-12 at 12 29 18 PM copy](https://github.com/user-attachments/assets/1e5e1cfe-325b-4e33-9bf5-d0012f478ea3)

**How to reproduce**
* Which ClickHouse server version to use: `24.6`

**Comment by jsc0218** [comment]
Could you provide a minimum reproducible example, as [mine with final=0](https://fiddle.clickhouse.com/38eee564-c75c-441d-ba2d-05447da95e23) here doesn't have the problem. It looks like it involves some opt of view.

**Comment by goums** [comment]
I also agree that the way `force_optimize_projection` and `force_optimize_projection_name` work is not fully satisfying, as it seems there is no way to actually force a query use a specific projection. 

`force_data_skipping_indices` behavior seems more explicit in comparison for example.
